File: 2022/day07/d7.php
<?php

declare(strict_types=1);

class Folder implements IteratorAggregate
{
    private array $files = [];

    public function __construct(
        public readonly ?Folder $parent = null
    ) {
    }

    public function addFolder(string $name): Folder
    {
        if (!isset($this->files[$name])) {
            $this->files[$name] = new Folder($this);
        }

        return $this->files[$name];
    }

    public function addFile(string $name, int $size): void
    {
        $this->files[$name] = $size;
    }

    public function getSize(): int
    {
        return array_reduce($this->files, fn ($c, $f) => $c + ($f instanceof Folder ? $f->getSize() : $f));
    }

    public function getIterator(): Traversable
    {
        yield from $this->files;
    }
}

class RecursiveFoldeIterator implements IteratorAggregate
{
    public function __construct(private readonly Folder $folder)
    {
    }

    public function getIterator(): Iterator
    {
        yield $this->folder;
        foreach ($this->folder as $key => $file) {
            if ($file instanceof Folder) {
                yield from new RecursiveFoldeIterator($file);
            }
        }
    }
}

$cwd = $root = new Folder();
$input = explode(PHP_EOL, trim(file_get_contents('php://stdin')));
foreach ($input as $line) {
    $tokens = explode(' ', $line);
    if ('$' == $tokens[0] && 'cd' == $tokens[1]) {
        $cwd = match ($tokens[2]) {
            '/' => $root,
            '..' => $cwd?->parent ?? $root,
            default => $cwd->addFolder($tokens[2]),
        };
    } elseif ('$' != $tokens[0]) {
        if ('dir' == $tokens[0]) {
            $cwd->addFolder($tokens[1]);
        } else {
            $cwd->addFile($tokens[1], (int) $tokens[0]);
        }
    }
}

$res1 = 0;
$res2 = 70000000;
$requiredSize = 30000000 - (70000000 - $root->getSize());

foreach (new RecursiveFoldeIterator($root) as $name => $folder) {
    if ($folder->getSize() < 100000) {
        $res1 += $folder->getSize();
    }
    if ($folder->getSize() > $requiredSize) {
        $res2 = min($res2, $folder->getSize());
    }
}

printf("Result 1: %d\nResult 2: %d\n", $res1, $res2);

File: 2022/day07/d7.cpp
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

class Directory {
  string name;
  Directory *parent;
  map<string, Directory *> children;
  map<string, int> files;

public:
  Directory(string name, Directory *parent = nullptr)
      : name(name), parent(parent) {}

  ~Directory() {
    for (const auto &child : children) {
      delete child.second;
    }
  }

  Directory *parent_dir() { return parent; }
  void createDirectory(const string name) {
    children[name] = new Directory(name, this);
  }
  void addFile(const string name, const unsigned int size) {
    files[name] = size;
  }
  Directory *getDirectory(const string name) const { return children.at(name); }

  vector<Directory *> getDirectories() const;

  unsigned int getSize() const;
};

vector<Directory *> Directory::getDirectories() const {
  vector<Directory *> directories;
  for (const auto &child : children) {
    directories.push_back(child.second);
    for (const auto &directory : child.second->getDirectories()) {
      directories.push_back(directory);
    }
  }
  return directories;
}

unsigned int Directory::getSize() const {
  unsigned int size{0};
  for (const auto &child : children) {
    size += child.second->getSize();
  }
  for (const auto &file : files) {
    size += file.second;
  }

  return size;
}

unsigned int part1(const Directory &root) {
  unsigned int result{0};
  for (const auto &directory : root.getDirectories()) {
    size_t size{directory->getSize()};
    if (size < 100000) {
      result += size;
    }
  }
  return result;
}

unsigned int part2(const Directory &root) {
  unsigned int requiredSize{30000000 - (70000000 - root.getSize())};
  unsigned int smallest{70000000};

  for (const auto &directory : root.getDirectories()) {
    unsigned int size{directory->getSize()};
    if (size >= requiredSize && size < smallest) {
      smallest = size;
    }
  }

  return smallest;
}

int main() {

  Directory root{"root"};
  Directory *cwd = &root;

  string line;
  while (getline(cin, line)) {
    if (line[0] == '$') {
      if (line.substr(0, 5) == "$ cd ") {
        string dir = line.substr(5);
        if (dir == "/") {
          cwd = &root;
        } else if (dir == "..") {
          cwd = cwd->parent_dir();
        } else {
          cwd = cwd->getDirectory(dir);
        }
      }
    } else if (line.substr(0, 4) == "dir ") {
      cwd->createDirectory(line.substr(4));
    } else {
      stringstream ss{line};
      unsigned int size;
      string filename;
      ss >> size >> filename;
      cwd->addFile(filename, size);
    }
  }

  cout << "Result 1: " << part1(root) << endl;
  cout << "Result 2: " << part2(root) << endl;
}
